[AMDGPU]Optimize SGPR spills

vg0204 opened the pull request:
This PR is dependent on [#93779](https://github.com/llvm/llvm-project/pull/93779).

As currently, each SGPR Spills are lowered to go into distinct stack slots in stack frame after SGPR allocation phase. Therefore, this patch utilizes the capability of StackSlotColoring to ensure the stack slot sharing if possible for stack frame index, where the SGPR spills are occuring in the non-interfering region.

StackSlotColoring is introduced immediately after SGPR register allocation, just to ensure that any further lowering happens on the optimally allocated stack slots, with certain flags to indicate the preservation of certain analysis result later to be used by RA of other register classes.

vg0204 committed:
[WIP] Implemented a patch to optimize SGPR spills.

Introduced the StackSlotColoring pass after SGPR RegAlloc and Spill to optimize stack slots reusage.

vg0204 committed:
Added configuration flag for StackSlotColoring & refactored comments.

arsenm commented on llvm/lib/CodeGen/StackSlotColoring.cpp:
```diff
@@ -67,6 +67,9 @@ namespace {
     SlotIndexes *Indexes = nullptr;
 
+    // - preserves Analysis passes in case RA may be called afterwards.
+    bool preserveRegAllocNeededAnalysis = false;
```
There shouldn't be a reason to have a configuration flag for this. The pass can just unconditionally preserve the analyses.

This should also be a separate PR, not just a separate commit in this PR

vg0204 replied:
> There shouldn't be a reason to have a configuration flag for this. The pass can just unconditionally preserve the analyses.
> 
> This should also be a separate PR, not just a separate commit in this PR

ok sure, I should put changes in generic pass as separate PR.

But, as the same stackSlotColoring pass exists in generic way(after RA) for other Target's pipeline where these analysis are not really required to be preserved. As in this case, it just got additionally called in between AMDGPU target pipeline(where addtional analysis needs to be preserved). Therefore, got suggested to introduce this flag to avoid avoid any changes for other architecture's pipeline( maybe caused due to additional pass preserving in stackSlotColoring).

arsenm replied:
Preserved does not mean required. Make them unconditionally preserved

cdevadas commented on llvm/test/CodeGen/AMDGPU/stack-slot-color-after-sgpr-alloc-equal-size-stack-objects.mir:
```diff
@@ -0,0 +1,127 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -verify-machineinstrs -stress-regalloc=3 -start-before=greedy -stop-before=prologepilog -o - %s | FileCheck -check-prefix=SHARE %s
```
-mcpu missing. in all new tests.
Since the RUN lines are the same combine all 3 tests into a single .mir file.

vg0204 replied:
> -mcpu missing. in all new tests. Since the RUN lines are the same combine all 3 tests into a single .mir file.

Can you enlighten me on what cpu to mention, as I am not sure exactly how to decide on.

Also, kindly comment on the configuration flag related changes.

cdevadas reviewed (commented):
Yes, the generic changes should go in a separate patch.

vg0204 committed:
[CodeGen] Preserved additional analyses in StackSlotColoring pass.

The pass pipeline of some architecture splits register allocation
phase based on different register classes. As some analyses need to
be computed at the beginning of the register allocation and kept
alive till all values are assigned to some physical registers.

This poses challenge with objective of introducing StackSlotColoring
after partial virtual registers are assigned to physical registers,
in order to optimize stack slots usage.As this pass doesn't preserve
few analysis yet to be needed by the register allocation of the
remaining virtual registers, necessiating them to be kept preserved.

jayfoad commented:
> [AMDGPU] Implemented a patch to optimize SGPR spills

Please keep subject lines concise. "Implemented a patch" just wastes words. If you remove that part then you can explain a little bit more about the optimization.

vg0204 commented:
> > [AMDGPU] Implemented a patch to optimize SGPR spills
> 
> Please keep subject lines concise. "Implemented a patch" just wastes words. If you remove that part then you can explain a little bit more about the optimization.

Point Noted!

vg0204 committed:
Made the following suggested changes.

Raised another PR itself to deal with generic changes in
StackSlotColoring pass needed to resolve current ticket.

Merged multiple mir test cases into one to be concise.

cdevadas commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1406,6 +1406,12 @@ bool GCNPassConfig::addRegAssignAndRewriteOptimized() {
   // since FastRegAlloc does the replacements itself.
   addPass(createVirtRegRewriter(false));
 
+  // As by this point SGPR's RA is done introducing SGPR spills to stack frame
```
```suggestion
  // As by this point SGPR's RA is done introducing SGPR spills to stack frame
  // through SGPRAllocPass. So, invoking StackSlotColoring here, may allow these
  // SGPR spills to re-use stack slots, before these spills is further lowered
  // down via SILowerSGPRSpills(i.e. equivalent of PEI for SGPRs).
```
 // At this point, the sgpr-regalloc has been done and it is good to have the stack
  // slot coloring to try to optimize the SGPR spill stack indices before attempting
  // the custom SGPR spill lowering.

cdevadas commented on llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:
```diff
@@ -278,15 +278,16 @@ void SILowerSGPRSpills::extendWWMVirtRegLiveness(MachineFunction &MF,
   for (auto Reg : MFI->getSGPRSpillVGPRs()) {
     for (MachineBasicBlock *SaveBlock : SaveBlocks) {
       MachineBasicBlock::iterator InsertBefore = SaveBlock->begin();
+
```
Extra new line.

cdevadas commented on llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:
```diff
@@ -278,15 +278,16 @@ void SILowerSGPRSpills::extendWWMVirtRegLiveness(MachineFunction &MF,
       // Set SGPR_SPILL asm printer flag
       MIB->setAsmPrinterFlag(AMDGPU::SGPR_SPILL);
-      if (LIS) {
+
```
Ditto.

cdevadas commented on llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:
```diff
@@ -300,6 +301,7 @@ void SILowerSGPRSpills::extendWWMVirtRegLiveness(MachineFunction &MF,
       auto MIB = BuildMI(*RestoreBlock, InsertBefore, DL,
                          TII->get(TargetOpcode::KILL));
       MIB.addReg(Reg);
+
```
Ditto.

cdevadas commented on llvm/test/CodeGen/AMDGPU/stack-slot-color-after-sgpr-regalloc.mir:
```diff
@@ -0,0 +1,353 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
```
Should precommit the test. But I'm not sure this test currently does anything significant to the slot optimization.
Add another run line to show the output after stack-slot-coloring as well to see the effect of this pass.
Should add a short description at each test what it meant for?
I don't see the slot-indices being shared for the SGPR spills.

vg0204 replied:
I believe I added the RUN command for both BEFORE & AFTER stack slot coloring pass (the one invoked just after sgpr regAlloc), & stack indices re-usage can be seen.

cdevadas reviewed (commented):
This patch needs a pre-commit PSDB test.

cdevadas commented on llvm/test/CodeGen/AMDGPU/spill-scavenge-offset.ll:
```diff
@@ -10098,7 +10098,7 @@ define amdgpu_kernel void @test_limited_sgpr(ptr addrspace(1) %out, ptr addrspac
 ; GFX6-NEXT:    v_lshlrev_b32_e32 v5, 8, v0
 ; GFX6-NEXT:    buffer_load_dwordx4 v[0:3], v[5:6], s[4:7], 0 addr64 offset:240
-; GFX6-NEXT:    s_mov_b32 s2, 0x84400
+; GFX6-NEXT:    s_mov_b32 s2, 0x86a00
```
Apart from the offset change, I don't see any difference in this test. Can you see why the offset is high?

vg0204 replied:
I found it. As when you call StackSlotColoring pass eventhough no stack slot usage is done, it also effectively re-assigns the FI to allocated stack slot. As a result, during PEI when actual offset is assigned they also differ. In this patch, the additional invocation of StackSlotColoring pass before SILowerSGPRSpills is just doing re-assignment of FI (with no effective slot re-usage) resulting in different offset assignment. But in overall, no change in stack memory usage (as no scope found for this optimization in this example).

cdevadas replied:
I believe, there is only the offset difference. The final stack consumption remains the same as earlier for the function.

vg0204 replied:
> I believe, there is only the offset difference. The final stack consumption remains the same as earlier for the function.

Yes

vg0204 commented on llvm/include/llvm/CodeGen/Passes.h:
```diff
@@ -283,7 +283,7 @@ namespace llvm {
   extern char &MachineCombinerID;
 
-  /// StackSlotColoring - This pass performs stack coloring and merging.
+  /// StackColoring - This pass performs stack coloring and merging.
```
> Why this change?

Need to get rid of it!

vg0204 committed:
Made changes to get rid of generic changes & addressed the required changes in comments and test files.

jayfoad commented on llvm/include/llvm/CodeGen/Passes.h:
```diff
@@ -283,7 +283,7 @@ namespace llvm {
   extern char &MachineCombinerID;
 
-  /// StackColoring - This pass performs stack coloring and merging.
+  /// StackSlotColoring - This pass performs stack coloring and merging.
```
Remove this. StackColoring and StackSlotColoring are two different passes.

jayfoad replied:
Sorry I think I was commenting on an old version of your patch.

cdevadas commented on llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:
```diff
@@ -1775,8 +1775,10 @@ bool SIRegisterInfo::spillSGPR(MachineBasicBlock::iterator MI, int Index,
-    assert(SB.NumSubRegs == VGPRSpills.size() &&
-           "Num of VGPR lanes should be equal to num of SGPRs spilled");
+    assert(SB.NumSubRegs <= VGPRSpills.size() &&
+           "Num of VGPR lanes should be greater or equal to num of SGPRs "
```
Keep this string as concise as possible. Any extra details you want to add should be included in a comment statement before the assert. The original string looks good. You might want to change "should be equal to" -> "should be less or equal to".

vg0204 replied:
I think that number of VGPR lanes corresponds to spill slot size(equal to largest spill stack object residing in it) which should be equal to or greater than number of SGPR spills created which corresponds to the spill stack object size.

vg0204 committed:
Remove unwanted changes not needed in this PR, and refactored some comments.

cdevadas commented on llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:
```diff
@@ -1775,10 +1775,13 @@ bool SIRegisterInfo::spillSGPR(MachineBasicBlock::iterator MI, int Index,
   if (SpillToVGPR) {
 
+    // Since stack slot coloring pass is trying to optimize SGPR spills,
+    // VGPR lanes (mapped from spill stack slot) may be shared for unequal SGPR
```
Change unequal SGPR spills into SGPR spills of different sizes.

cdevadas commented on llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:
```diff
@@ -1775,10 +1775,13 @@ bool SIRegisterInfo::spillSGPR(MachineBasicBlock::iterator MI, int Index,
-           "Num of VGPR lanes should be greater or equal to num of SGPRs "
-           "spilled, as Stack Slot Coloring pass assigns different SGPR spills "
-           "into same stack slots");
+           "Num of VGPR lanes mapped should be greater or equal to num of "
```
I guess it is more appropriate to use 'shouldn't be less than' instead of 'should be greater or equal'

vg0204 replied:
> I guess it is more appropriate to use 'shouldn't be less than' instead of 'should be greater or equal'

Then should I reverse the assert statement, so that 'shouldn't be less than' can be used?

cdevadas reviewed (commented):
The new test llvm/test/CodeGen/AMDGPU/stack-slot-color-after-sgpr-regalloc.mir needs to be written precisely to show the stack indices being shared. See my previous comments.

vg0204 commented:
> The new test llvm/test/CodeGen/AMDGPU/stack-slot-color-after-sgpr-regalloc.mir needs to be written precisely to show the stack indices being shared. See my previous comments.

While going through it, I found a problem in test case output, need to reassess my changes in PR, currently working on it.

vg0204 commented:
> > The new test llvm/test/CodeGen/AMDGPU/stack-slot-color-after-sgpr-regalloc.mir needs to be written precisely to show the stack indices being shared. See my previous comments.
> 
> While going through it, I found a problem in test case output, need to reassess my changes in PR, currently working on it.

Everything is alright! no changes is needed

vg0204 committed:
Made comment changes in SIRegisterInfo.cpp.

vg0204 committed:
Updated the  MIR LIT showing the VGPR lane usage during SGPR spills.

cdevadas commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -0,0 +1,295 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -verify-machineinstrs -stress-regalloc=3 -start-before=greedy -stop-after=stack-slot-coloring -o - %s | FileCheck -check-prefix=CHECK %s
```
Should use a different check-prefix instead of the default CHECK. May be STACK-SLOT-OPT?

cdevadas replied:
There are now two instances of stack-slot-coloring in our pipeline. You're interested in the first invocation.  Better specify it with the index too, like -stop-after=stack-slot-coloring,0. The index zero here is meant for stopping after the first instance.

cdevadas commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -0,0 +1,295 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
```
Test filename can be optimize-vgpr-lanes-for-sgpr-spills.mir?

vg0204 replied:
But, I think this filename is as good after patch is included, here it is not done yet, right?

arsenm commented on llvm/test/CodeGen/AMDGPU/llc-pipeline.ll:
```diff
@@ -362,10 +362,12 @@
 ; GCN-O1-NEXT:        Virtual Register Map
 ; GCN-O1-NEXT:        Live Register Matrix
 ; GCN-O1-NEXT:        SI Pre-allocate WWM Registers
+; GCN-O1-NEXT:        Live Stack Slot Analysis
```
Can we keep this preserved in the previous passes?

vg0204 replied:
I tried this & found that as StackSlotColoring pass is not currently updating LiveStacks(LS) analysis, eventhough modifying the stack slots usage (as before it is being invoked only after complete RA when all spills happened just before PEI, so LS not needed anymore), so LS results are invalidated rather preserved. But, now LS is being needed by RA (for remaining register class) later, so LS analysis appears later as being re-calculated.

arsenm replied:
Right, can the pass be updated so that it can be preserved?

vg0204 replied:
Need to look into it, but I cannot say how much changes might it take. Also, as the same LiveStack(LS) would be propagated into RA pass later for entring additional entries, RA might have some side effects while working on presereved LS. Considering all this, this updates might take big significant effort, followed by rigourous tests, so nothing else breaks.

vg0204 replied:
@arsenm, after spending some time & tinkering around, I figured it out! Let me know so the Stack Slot Coloring should preserve LiveStack analysis?

arsenm replied:
Yes, open a follow up PR to do that

vg0204 replied:
As there already exists a PR #93779, which deals with preserving additional analysis to be later used by RA passes, can I incorporate these changes there OR open a new follow-up PR?

arsenm replied:
You should nearly always break it into the smallest possible separable pieces

jayfoad replied:
@vg0204 ping: are you planning to make StackSlotColoring preserve LiveStacks, or is there some reason it is hard to do?

vg0204 committed:
Update LIT mir test case with optimized SGPR spill.

cdevadas reviewed (approved):
You don't have to wait for every reviewer to approve the patch unless one of the reviewers mentions it.

jplehr commented:
Hi, I believe this turned one of our buildbots red: https://lab.llvm.org/buildbot/#/builders/123/builds/253
If you require assistance in reproducing, let me know and I'll see if I can be of help.

vg0204 commented:
> Hi, I believe this turned one of our buildbots red: https://lab.llvm.org/buildbot/#/builders/123/builds/253 If you require assistance in reproducing, let me know and I'll see if I can be of help.

Yeah, that would be great!

yxsamliu commented:
> > Hi, I believe this turned one of our buildbots red: https://lab.llvm.org/buildbot/#/builders/123/builds/253 If you require assistance in reproducing, let me know and I'll see if I can be of help.
> 
> Yeah, that would be great!

I can confirm that this PR caused regression in blender test of buildbot:

```
Fra:48 Mem:1531.14M (Peak 1531.14M) | Time:06:58.63 | Mem:2448.05M, Peak:2448.05M | Scene, View Layer | Sample 0/256
Warning: precise memory violation signal reporting is not enabled, reported
location may not be accurate.  See "show amdgpu precise-memory".

Thread 103 "blender" received signal SIGSEGV, Segmentation fault.
[Switching to thread 103, lane 9 (AMDGPU Lane 1:1:1:1/9 (0,0,0)[9,0,0])]
0x00007ffd7689a68c in integrator_intersect_shadow(KernelGlobalsGPU const*, int) () from memory://192866#offset=0x7ffd77f01240&size=4555968
```

Currently the buildbot has disabled all Blender test scenes so it is apparently green, but we'd get this fixed ASAP since this commit may get to amd-staging branch and cause other regressions.

debugger shows memfault happens at the highlighted line:

<img width="936" alt="Capture" src="https://github.com/llvm/llvm-project/assets/6145511/461c848d-2e8a-4aa2-ba88-8065f63ce77a">

the real instruction causing memfault is most likely the `buffer_load_dword` right above it, where `s[0:3]` contains unreasonably large values.

yxsamliu commented:
[kernel-hip-amdgcn-amd-amdhsa-gfx908.zip](https://github.com/user-attachments/files/15905813/kernel-hip-amdgcn-amd-amdhsa-gfx908.zip)

To accurately reproduce the assembly file, use the following clang command:

```
clang -cc1 -mcode-object-version=4 -mllvm --amdhsa-code-object-version=4 -triple amdgcn-amd-amdhsa -aux-triple x86_64-unknown-linux-gnu -S -disable-free -mrelocation-model pic -pic-level 2 -mframe-pointer=none -aux-target-cpu x86-64 -mllvm -amdgpu-internalize-symbols -target-cpu gfx908 -O3 -vectorize-loops -vectorize-slp -faddrsig -o tmp.s -x ir kernel-hip-amdgcn-amd-amdhsa-gfx908.bc
```

vg0204 commented:
> [kernel-hip-amdgcn-amd-amdhsa-gfx908.zip](https://github.com/user-attachments/files/15905813/kernel-hip-amdgcn-amd-amdhsa-gfx908.zip)
> 
> To accurately reproduce the assembly file, use the following clang command:
> 
> ```
> clang -cc1 -mcode-object-version=4 -mllvm --amdhsa-code-object-version=4 -triple amdgcn-amd-amdhsa -aux-triple x86_64-unknown-linux-gnu -S -disable-free -mrelocation-model pic -pic-level 2 -mframe-pointer=none -aux-target-cpu x86-64 -mllvm -amdgpu-internalize-symbols -target-cpu gfx908 -O3 -vectorize-loops -vectorize-slp -faddrsig -o tmp.s -x ir kernel-hip-amdgcn-amd-amdhsa-gfx908.bc
> ```

Getting this while trying to reproduce.

![dumpError](https://github.com/llvm/llvm-project/assets/35700483/f09aa238-13a4-4fd9-b570-0d0606096138)

arsenm commented:
> > [kernel-hip-amdgcn-amd-amdhsa-gfx908.zip](https://github.com/user-attachments/files/15905813/kernel-hip-amdgcn-amd-amdhsa-gfx908.zip)
> 
> > 
> 
> > To accurately reproduce the assembly file, use the following clang command:
> 
> > 
> 
> > ```
> 
> > clang -cc1 -mcode-object-version=4 -mllvm --amdhsa-code-object-version=4 -triple amdgcn-amd-amdhsa -aux-triple x86_64-unknown-linux-gnu -S -disable-free -mrelocation-model pic -pic-level 2 -mframe-pointer=none -aux-target-cpu x86-64 -mllvm -amdgpu-internalize-symbols -target-cpu gfx908 -O3 -vectorize-loops -vectorize-slp -faddrsig -o tmp.s -x ir kernel-hip-amdgcn-amd-amdhsa-gfx908.bc
> 
> > ```
> 
> 
> 
> Getting this while trying to reproduce.
> 
> 
> 
> ![dumpError](https://github.com/llvm/llvm-project/assets/35700483/f09aa238-13a4-4fd9-b570-0d0606096138)
> 
> 

This is due to mixing upstream and amd-staging with the different debug info

arsenm commented:
> Can you please elaborate?


See https://github.com/llvm/llvm-project/issues/56194

yxsamliu commented:
let me try getting a bitcode without debug info

yxsamliu commented:
[res.new.zip](https://github.com/user-attachments/files/15914966/res.new.zip)
bc and assembly without debug info are attached. pls use 4b9112e88a998ce620e4683548f2afd17cc5fe95 clang and the above mentioned command to get assembly and compare with the attached one to make sure they are the same.

vg0204 commented:
Also, I do noat have any clue how to proceed with this bitcode & assembly files provided by you @yxsamliu , to help in fixing the blender test caused by the commit. So if you could help me out in that, it would be great!

yxsamliu commented:
the idea is to locate the buffer_load_dword in  integrator_intersect_shadow which has invalid s[0:3] in the bad case and locate the counter part in the good case. By comparing the ISA, find the def of s[0:3] or chase further to find the instruction that causes invalid def of s[0:3]. Then find the invalid transformation that causes that instruction by focusing on how each pass transform that instruction.

If the ISA of integrator_intersect_shadow is too different in the bad/good cases we can force noinline for callees of integrator_intersect_shadow in the hope of narrow down the issue in its callees. We may also consider other ways to minimize the difference between the bad/good cases, e.g. adding some counter or threshold in the newly added pass to control the number of changes it makes to pinpoint the minimal change that causes pass/fail.

yxsamliu commented:
Can we temporarily revert this patch while we investigate the regression? Thanks.

cdevadas commented:
> Can we temporarily revert this patch while we investigate the regression? Thanks.

Yes, revert it to unblock the blender testing.

vg0204 commented:
> the idea is to locate the buffer_load_dword in integrator_intersect_shadow which has invalid s[0:3] in the bad case and locate the counter part in the good case. By comparing the ISA, find the def of s[0:3] or chase further to find the instruction that causes invalid def of s[0:3]. Then find the invalid transformation that causes that instruction by focusing on how each pass transform that instruction.
> 
> If the ISA of integrator_intersect_shadow is too different in the bad/good cases we can force noinline for callees of integrator_intersect_shadow in the hope of narrow down the issue in its callees. We may also consider other ways to minimize the difference between the bad/good cases, e.g. adding some counter or threshold in the newly added pass to control the number of changes it makes to pinpoint the minimal change that causes pass/fail.

Thank you Sam for this heads up!
